src/lib/core/tt_hostname.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+
+enum {
+	TT_HOSTNAME_LEN_MAX = 63,
+	TT_HOSTNAME_SIZE_MAX = TT_HOSTNAME_LEN_MAX + 1,
+};
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/**
+ * DNS- and host-friendly name. Only lowercase letters, digits, '-', starts with
+ * a letter, max 63 chars not including terminating zero.
+ */
+struct tt_hostname {
+	/** Name characters. */
+	char data[TT_HOSTNAME_SIZE_MAX];
+};
```

Comment:
TBH I don't see any point in this abstraction. Basically, it's a fixed-length string. Why do we need to wrap it in a struct? We don't do that for space names, for example, and it's unclear why host names should be different. I think we should use a plain string instead. If we want to allocate it statically, we can introduce a constant for the max length (`REPLICA_NAME_MAX`). And to check a name validity we could introduce a helper function, similar to `identifier_check`.

A few comments regarding the API:
1. The API name (`tt_hostname`) is rather confusing because it's not about host names - it's about Tarantool instance/replicas names, which use different naming rules than host names.
2. It's confusing that some `tt_hostname_` methods work with plain strings.
3. Some methods are purely auxiliary and never used outside `tt_hostname.h` (e.g. `tt_hostname_safelen`), but still they are defined like a part of the public API.
4. `tt_hostname_for_log` looks useless - you could just as well teach `tt_hostname_str` return `"<no-name>"` if the argument is nil.
5. `tt_hostname_are_eq` - we usually name functions like that `FOO_is_equal` AFAICS.

Reply:
I don't mind changing it back to a plain string (it was implemented as a plain string in the first version). Although I think it is better to stay encapsulated like now, because the rules are much stricter than for space names or any other names.

Besides, space names you can't store in a struct conveniently. They are always `const char *`, because the limit is 65k. You can't possibly store such static buffer in each space and space def.

With the hostnames the limit is just 64 bytes. Which means they are stored as a static buffer everywhere. It is just shorter to define them as a static buf one time inside of a struct and then use this struct.

If that is not persuasive enough, you need to discuss it with @sergepetrenko. He also was for the struct-version. If you both agree on the plain string, then I will change it.

I can also address a couple of points right away while you two decide:

- *"use different naming rules than host names"* - that isn't true, the rules are the same: https://man7.org/linux/man-pages/man7/hostname.7.html#:~:text=Each%20element%20of%20the%20hostname,not%20start%20with%20a%20hyphen..

- *"Some methods are purely auxiliary "* - and what? Should I move one-liners into the .c file just because they use some other one-liner? I don't see it as a problem. On the contrary, it allows to have less code. I would prefer to keep it as is (unless it will become not needed in the plain string version of the names).

Reply:
Better refer to [RFC](https://www.rfc-editor.org/rfc/rfc1035#section-2.3.1) than to man

Reply:
> With the hostnames the limit is just 64 bytes. Which means they are stored as a static buffer everywhere. It is just shorter to define them as a static buf one time inside of a struct and then use this struct.

So this looks like a limitation of the programming language. I assume if we wrote in C++, we wouldn't even think of introducing a wrapper like this - instead we'd use `std::string` directly. That's why `tt_hostname` abstraction looks artificial to me. Maybe, we could add a fixed-length string struct to the core lib instead? (Not sure if it's worth it though.)

> If that is not persuasive enough, you need to discuss it with @sergepetrenko. He also was for the struct-version. If you both agree on the plain string, then I will change it.

OK, I'll discuss it with @alyapunov and @sergepetrenko.

> "use different naming rules than host names" - that isn't true, the rules are the same

The document you refer to says that a hostname may be up to 253 characters long and contain dots. The length of `tt_hostname` is limited by 63 characters and it can't contain dots. That is, it's not a host name, it's a sub domain name.

> * _"Some methods are purely auxiliary "_ - and what? Should I move one-liners into the .c file just because they use some other one-liner?

No, moving them to the C file isn't a good idea. I'd rather you inline most of them (e.g. `tt_hostname_safelen` doesn't really make the code shorter or easier to follow IMO). Also, you could add the `_internal` suffix or group internal methods together so that it's clear which methods are public and which are private and shouldn't be called by the API users.

Reply:
Just for the record, here's my original comment where I asked @Gerold103 to redo this with `struct tt_hostname` instead of a plain string: https://github.com/tarantool/tarantool/pull/8289#discussion_r1108445150.

>> With the hostnames the limit is just 64 bytes. Which means they are stored as a static buffer everywhere. It is just shorter to define them as a static buf one time inside of a struct and then use this struct.

> So this looks like a limitation of the programming language. I assume if we wrote in C++, we wouldn't even think of introducing a wrapper like this - instead we'd use `std::string` directly. That's why `tt_hostname` abstraction looks artificial to me. Maybe, we could add a fixed-length string struct to the core lib instead? (Not sure if it's worth it though.)

IMO the main reason to use `struct tt_hostname` instead of a plain string is not this, but rather the importance of validation:
tt_hostname isn't just any string, but a quite specific one, and one of a specific length.

Yes, you may remember to call the validation method every time and copy the string even without this abstraction, but having `tt_hostname_set_str` do that for you feels more convenient.

BTW, if we agree to leave `tt_hostname` approach, I suggest to merge `tt_hostname_str_is_valid_n` and `tt_hostname_set_str_n`. The latter should perform the validation and return the error if needed (@Gerold103).

Reply:
> The document you refer to says that a hostname may be up to 253 characters long and contain dots

Sorry, I wasn't attentive enough. I saw the phrase `from 1 to 63 characters long` and somewhy thought it relates to the entire name, not its part.

> `tt_hostname_safelen` doesn't really make the code shorter or easier to follow IMO

I think it does. Also IMO. If you try to inline `tt_hostname_safelen()`, you will see that it makes the lines notably longer. The less line-break wraps and `{}` I have, the easier I find the code to follow. If I would inline it, then `tt_hostname_set_str()` will get 2 pairs of `{}` and +1 line break.

Adding `_internal` in this case makes the functions useless. It won't be short anymore. I can only suggest to imagine it is not internal then. Just not used outside of the header "yet".

Reply:
> If that is not persuasive enough, you need to discuss it with @sergepetrenko. He also was for the struct-version. If you both agree on the plain string, then I will change it.

We didn't come to an agreement with @locker and @alyapunov, so let's continue here.

Here's my point (already stated a couple of comments above):

> IMO the main reason to use struct tt_hostname instead of a plain string is not this, but rather the importance of validation: tt_hostname isn't just any string, but a quite specific one, and one of a specific length.

I'll elaborate on that a bit:

imagine you have
```c
char name[64];
```
It's just a string, and everywhere where you expect a "hostname" you will call
```c
tt_hostname_validate(str);
strcpy(name, str, strlen(str));
```
It's quite possible to forget the validate() routine by accident. So it'd be nice to have a single method
```c
int
tt_hostname_set(char *name, const char *str);
```
which performs validation and copy in one step.

Now, if the out-parameter of `tt_hostname_set` is a plain string, like @locker and @alyapunov suggest, again, there's nothing which would **force** the programmer to operate on it correctly. Someone might as well try to set a name by calling `strcpy` or whatever.

This is where a separate struct comes in handy: imagine you have
```c
struct tt_hostname {
    char buf[64];
};

int
tt_hostname_set(struct tt_hostname *name, const char *str);
```

Since this is the ony constructor for `tt_hostname` you can' end up with an invalid name. And it's really hard to bypass. I can't imagine anyone doing this instead of using a constructor:
```c
strcpy(name.buf, str, 64);
```
It takes much more effort to write such erroneous code compared to when `name` was just a string.

I used `64` instead of `TT_HOSTNAME_SIZE_MAX` in this example just for simplicity.

To sum up: I'm for leaving this a separate struct. It represents a particular concept (a subdomain name), and it helps to write correct code.

Reply:
Oh, a bike shed! I also have an opinion that I must represent.

As far as I understood `tt_hostname` class is designed for storing a fixed length string with a special validation. The idea of that class is that if an object stores a string, it is guaranteed to be validated by object's constructor. The implementation is simple: we imaginarily hide the string inside of `tt_hostname` structure and give a word that we will not access it directly, only via special methods.

The problem I see is that the class has to implement all generic string methods, like `are_equal` etc. And, for example, if we need a hash table of hostnames, we will need to implement a new method like `tt_hostname_crc32` or `tt_hostname_murmur`. At some point we will need `tt_hostname_cmp` etc.

If we consider that OK, what will be next? If we want to use the same approach for (for instance) for space name (`struct tt_space_name`), will will need to implement tons of methods again, `tt_space_name_cmp`, `tt_space_name_hash` etc. Of cause we will use some kind of dynamic string inside of this new structure but we will still pretend not to access it directly.

I think our purpose is to make things more simple.

I would like to have some generic string implementations, both for fixed size and for dynamic size. In C that would be inevitably a bit ugly, but still possible. Imagine code generating macros like `#define FIXED_STRING_GEN(name, max_length)` and `#define DYNAMIC_STRING_GEN(name, no_arg)` that would generate `struct name` with lots of generic methods like:
```c
int name_set(struct name *s, const char *str);
int name_setl(struct name *s, const char *str, size_t len);
int name_copy(struct name *dst, const struct name *src);
void name_xcopy(struct name *dst, const struct name *src);
void name_move(struct name *dst, struct name *src);
...
int name_cmp(const struct name *s1, const struct name *s2);
const char *name_data(const struct name *s);
size_t name_size(const struct name *s);
const char *name_cstr(const struct name *s);
```

That would be a lot of code but it would be a very useful class. Note that since both types of string has that same API we can simply replace one implementation with another. Even more, we could create some more complex string wrappers, for example a class that is tolerant to passing a null object, something like `#define NULLABLE_STRING_GEN(name, base_string_macro, base_arg)`. And finally make a wrapper that would validate a given string with given function: "#define VALIDATED_STRING_GEN(name, validate_f, base_string_macro, base_arg)". That would be a lot of code too, but it would be only generic code. And how cool would be to create types for particular cases!
```c
bool validate_subdomain_name(const char *);
enum {SUBDOMAN_NAME_MAX_LEN = 64, };

VALIDATED_STRING_GEN(subdomain_name, validate_subdomain_name, NULLABLE_STRING_GEN, 50);

struct subdomain_name INSTANCE_NAME_CFG; // Actually uppercase for non-compile-time-constants bothers me
```

Reply:
Sorry for many letters. Shortly:
that should be a subdomain name, not host name,
I would prefer that in tt_hostname.h (or similar) there was only one defined function `..._validate`,
in any case all non-public API methods should be hidden with special namespace or name suffix,
uppercase names are reserved for macro and known-at-compile-time constants,

Reply:
You have to come to an agreement. I am keeping it a struct until you (@sergepetrenko and @locker) decide on something. For now I will just rename it as it is obviously is not a hostname, my bad.

Reply:
> @sergepetrenko also didn't agree on that.

Thanks, @Gerold103, I was a bit tired of arguing already. I did agree on that as a possible compromise, but @locker declined it anyway. So the choice still stands between `struct tt_hostname` and `char *` only for the hostname. Not for all the other names.

Reply:
In the new version I
* Renamed the struct to `tt_node_name`;
* Deleted `..._str_is_valid_n()` like Sergey suggested, and `..._str_is_valid()` too. Although for that I had to allow null-name in `..._set_str()` and `..._set_str_n()`;
* `tt_node_name_for_log()` is deleted like Vova suggested. Now `tt_node_name_str()` returns always a non-empty name suitable for logs.

Reply:
As a chthonic evil I was awakened, but the awakening took that long time.

Before TL;DR: my humble contribution is 'make it plain string'.

I would try to put pros and cons on one page
- string's size and format validation `tt_node_name_set_str_n` returns the `tt_node_name` VS plain `char *`
- size in memory to store the string (struct + static buffer VS plain `char *`)
- uniform representation of strings of various types and formats across Tarantool sources as plain `char *` VS customization of all varieties (with possible automation of code [de]duplication)
- there was a point - can't find above in discussion - that making the name string 'hidden' in the struct may prevent from format abuse

To my understanding, none of the above is a good grounds for introducing a struct: there are no myriads of places the name is set - I found just two of them - and hardly will ever be changed in the future. No strong protection is possible from the format abuse and the memory overhead for 3 strings looks waaaay too penny wise and pound foolish on that long discussion.  

The only part left - readability and comprehensibility which doesn't gain from the wrapping name in a struct, while making code motlier. Yes, it is far from being an easy reading by now but should we keep making it more like this?

Reply:
It was done as a plain string at first and it looked way worse. Just look at all the places where the names are used in the logs - for an empty name I had a convenient function `tt_node_name_str(const struct tt_node_name *)` which returned a special stub for empty names. I will also still have to store somewhere the max len constant to be able to use it in places like `xrow.h` - it can't be in `replication.h` because xrow doesn't depend on replication, and I am not even sure where to put it then. I also don't know yet where will I put the validation functions now. Keeping tt_node_name library with just functions? I have no idea. But ok, I can redo it third time. Then maybe back when we see how ugly it is.

Reply:
> It was done as a plain string at first and it looked way worse. Just look at all the places where the names are used in the logs - for an empty name I had a convenient function `tt_node_name_str(const struct tt_node_name *)` which returned a special stub for empty names.

You can add a helper function that takes a plain string or `NULL` and returns a special stub name for `NULL`.

> I will also still have to store somewhere the max len constant to be able to use it in places like `xrow.h` - it can't be in `replication.h` because xrow doesn't depend on replication, and I am not even sure where to put it then. I also don't know yet where will I put the validation functions now. Keeping tt_node_name library with just functions? I have no idea.

You can put them in `src/box/instance_name.h` (similarly to `src/box/identifier.h`). These functions shouldn't be a part of the core library.

Also, I don't think you should copy an instance name to requests (`join_request`, `subscribe_request`, etc).  You can just put a pointer to the MsgPack string there, like we do with other requests (e.g. `watch_request::name`).

Reply:
I redid the patch to use raw strings.

> Also, I don't think you should copy an instance name to requests (join_request, subscribe_request, etc). You can just put a pointer to the MsgPack string there

This I didn't do. There are too many places where I have to compare those strings. Imagine carring len with each `tt_uuid` in xrow and in all compare and set functions - this has the same effect. I tried to do it this way and it looks incomprehensible.